### ISSUE TITLE
DEFINE-PIPE is not idempotent

### ISSUE THREAD

**Issue by phoe** [Shinmera/verbose#7]
```lisp
CL-USER> (ql:quickload :verbose)
To load "verbose":
  Load 1 ASDF system:
    verbose
; Loading "verbose"

(:VERBOSE)
CL-USER> (v:info '(:foo) "sdfsdf")
NIL
2021-03-27 14:41:07 [INFO ] <FOO>: sdfsdf
CL-USER> (verbose:define-pipe () (v:repl-faucet))
#<VERBOSE:CONTROLLER :threaded :running :queue-size 0>
CL-USER> (v:info '(:foo) "sdfsdf")
NIL
2021-03-27 14:41:15 [INFO ] <FOO>: sdfsdf
2021-03-27 14:41:15 [INFO ] <FOO>: sdfsdf
CL-USER> (verbose:define-pipe () (v:repl-faucet))
#<VERBOSE:CONTROLLER :threaded :running :queue-size 0>
CL-USER> (v:info '(:foo) "sdfsdf")
NIL
2021-03-27 14:41:19 [INFO ] <FOO>: sdfsdf
2021-03-27 14:41:19 [INFO ] <FOO>: sdfsdf
2021-03-27 14:41:19 [INFO ] <FOO>: sdfsdf
```

According to the principle of least surprise and consistency with `defun`, `defparameter`, `defclass` et cetera, I expected that a secondary call to `define-pipe` is going to explicitly redefine the previously defined pipe, instead of adding on top of it.

* Is this the expected behavior?
* If yes, what is the way to explicitly replace the previous pipe?

**Comment by Shinmera** [comment]
There's no name argument, so it can only add pipes, unfortunately. 

You can remove it with `(piping:remove-segment v:*global-controller* (list (1- (length (piping:pipeline v:*global-controller*)))))` which is much more of a mouthful than I'd like, so patches to piping to improve the usability would be appreciated.

**Comment by phoe** [comment]
> There's no name argument, so it can only add pipes, unfortunately.

There's an optional controller argument though; I expect that every controller has its own singular pipe, and that this is what `define-pipe` operates on, and that calling `define-pipe` causes it to be redefined rather than augmented.

Because of this, naming this operator `define-pipe` is a consistency bug with standard CL `deffoo`/`define-foo` operators, because it doesn't define the global pipe, it adds new stuff on top of it.

I'd suggest moving the current implementation into e.g. `augment-pipe` and changing `define-pipe` to actually define a pipe. Would you accept a PR that does this?

**Comment by Shinmera** [comment]
This library is old and there's bound to be code out there that depends on this behaviour. Hard to say how much this would break.

**Comment by Shinmera** [comment]
I would be more amenable to changing the syntax to be backwards compatible, but allow adding a name for the new pipe so it can be changed.

**Comment by phoe** [comment]
Nowadays I'm generally opposed to permanent backwards compatibility, even in CL. Even SBCL's trademark "minor incompatible changes" tend to break backwards compatibility now and then.

What kind of name would you imagine though? What would be the default name, and how would it be resolved? Currently the pipe seems to be anonymous; the only thing that we have in Verbose is the controller, hidden behind an exported dynamic variable.

**Comment by Shinmera** [comment]
The name would be for the pipe you define? Segments can already be named.

**Comment by phoe** [comment]
OK, let me rephrase. Currently, `(verbose:define-pipe () (v:repl-faucet))` adds stuff on top of the pipe of the current controller.

I would like a bit of Verbose code that actually *replaces* the whole pipe instead of adding onto it. What would it look like?

**Comment by Shinmera** [comment]
You are misunderstanding. The controller is a `pipeline` and as such has a vector of pipes to distribute to. `define-pipe` naturally defines a new pipe on this pipeline. If you want to replace the entire pipeline you're in the wrong place to begin with.

**Comment by Shinmera** [comment]
The easiest way to do what you want is to just throw away the old controller and make a new one that has the pipes you want in it, similar to how `make-standard-global-controller` does it.

**Comment by phoe** [comment]
OK - I think I understand the hierarchy a little bit better now. Seems that I'm looking for something that would replace the whole pipeline with one that we want.

Options I see: 
* `define-pipeline`
* `define-controller`
* `setf v:*global-controller*`
* `setf (piping:pipeline v:*global-controller*)`

I think this last one, along with `make-standard-global-controller`, would be the best pick, especially since they are already implemented. Would you accept a PR to README that describes how to use that to fix the original problem?

**Comment by Shinmera** [comment]
I would accept a PR to add a `define-controller` macro to do what you want in a more trivial fashion, along with README amends to document it. It would have to take care to stop the old controller (if present), so things aren't quite as simple as just setting the global controller variable.

**Comment by Shinmera** [comment]
Actually thinking about it, it would be better if it didn't remove the old controller if present, and instead used `reinitialize-instance` to supply a new vector of pipes to use. Not sure if it's currently set up to allow that, so that behaviour might need to be added as well.

**Comment by phoe** [comment]
OK, I think this is doable, I'll try to prepare and submit it. If I don't do it in a reasonable time frame, please mercilessly ping and mention me by all available means.

**Comment by Shinmera** [comment]
Since I'm just as forgetful you'll have to ping me to ping you ;)
